Use before? and after? (#165)
andycroll/andycroll.com: _posts/2022/2022-09-19-use-before-and-after-methods-when-comparing-dates-and-times-in-rails.md

---
layout: article
title: Use before? and after? Methods When Comparing Dates and Times in Rails

> description: I can’t remember which is greater than or less than either.

Reviewer: 😆

category: ruby
image:
  base: 2022/use-before-and-after-methods-when-comparing-dates-and-times-in-rails
  alt: Clocks piled on top of each other
  source: https://unsplash.com/photos/FlHdnPO6dlw
  credit: Jon Tyson
---

> Active Support contains many additions to the basic classes that form the standard libraries of Ruby. There are extensions to `String`, `Hash`, `Array` and even `Date` and `Time`.

Reviewer: This first sentence does feel a little long but I couldn't think of a way to shorten it first time around. Wonder if there's a more succinct way of saying all of this.

Author: Think it's fine.

> Active Support contains many additions to the basic classes that form the standard libraries of Ruby. There are extensions to `String`s, `Hash`es, `Array`s and even the classes that manage `Date`s and `Time`s.

Reviewer: I would just leave it as "...extensions to `String`, `Hash`, `Array`, and even `Date` and `Time`."

Author: ```suggestion
Active Support contains many additions to the basic classes that form the standard libraries of Ruby. There are extensions to `String`, `Hash`, `Array` and even `Date` and `Time`.
```

> If you’ve been working inside Rails you might not realise that these are Rails-isms rather than Ruby-isms.

Reviewer: *If you spend most of your time working inside Rails, you might not realise...

(But also, doesn't this sentence only work the other way around, if they were in fact 'Ruby-isms'. Like wouldn't you assume they were Rails-isms...? Maybe not...)

Author: ```suggestion
If you spend most of your time working inside Rails applications, you might not realise that these are Rails-isms rather than Ruby-isms.
```

> In many cases if you’ve mostly been working inside Rails you might not realise that these are Rails-isms rather than Ruby-isms.

Reviewer: ```suggestion
If you’ve been working inside Rails you might not realise that these are Rails-isms rather than Ruby-isms.
```

Not sure of the positioning of 'mostly' really...

Some folks dislike these ”superfluous” additions. Some of these additions end up back inside the Ruby standard library itself. _Some_ of these methods help me not make mistakes when comparing times and dates.

## Instead of...

...comparing dates and times with greater than or less than operators:

```ruby
Date.new(1979, 9, 12) > Time.zone.now
#=> false

10.minutes_ago < 5.minutes.from_now
#=> true
```

## Use...

> ...`before?` and `after?`.

Reviewer: I definitely need these in my life. 😆

```ruby
Date.new(1979, 9, 12).after?(Time.zone.now)
#=> false

10.minutes_ago.before?(5.minutes.from_now)
#=> true
```

## Why?

> This is a personal stylistic choice. I find the readability of the logic to be hugely improved when using the `before?` and `after?` methods. This helps me reason about the code as I'm writing it and, more importantly, to understand my logic when I come back to it months from now.

Reviewer: ```suggestion
This is a personal stylistic choice. I find the readability to be hugely improved when using `before?` and `after?`. It helps me to reason about the code much more easily as I'm writing it and, more importantly, to quickly understand my logic when I come back to it months later.
```
